File: QuipMac/Services/APNsClient.swift
import Foundation
import CryptoKit

/// Errors surfaced from `APNsClient.send(...)`. The distinct cases let
/// the Settings UI and the auto-remove path respond differently: an
/// `unregistered` token → drop from the device store, a `throttled` →
/// retry once, `badKey` → surface an inline error in the Settings pane.
enum APNsError: Error, Equatable {
    case missingKey
    case invalidKey(String)
    case badRequest(String)
    case unregistered      // 410 / BadDeviceToken / Unregistered — drop the device
    case throttled         // 429 / 503 — retry later
    case serverError(Int)  // 5xx other than 503
    case unknown(Int, String)
}

/// Token signed with the APNs auth key. APNs rejects JWTs older than 60
/// minutes, so we cache and rotate at 50min to leave headroom.
private struct CachedJWT {
    let token: String
    let issuedAt: Date
    var isExpired: Bool {
        Date().timeIntervalSince(issuedAt) > 50 * 60
    }
}

/// Signs ES256 JWTs with the .p8 key and POSTs them + the alert payload
/// to Apple's HTTP/2 APNs endpoint. URLSession on macOS 14+ auto-upgrades
/// to HTTP/2 when the server supports it, so we don't need a separate
/// HTTP/2 client library.
///
/// Usage (from QuipMacApp):
/// ```
/// let client = APNsClient(keyId: ..., teamId: ..., bundleId: ...)
/// try await client.send(payload: [...], toDevice: device)
/// ```
///
/// Actor so the cached JWT + send pipeline can be reached from
/// background tasks safely. Construction throws because we parse the
/// .p8 key up-front — constructing many short-lived clients is fine
/// (one ES256 init is cheap; the JWT sign is the only expensive op).
actor APNsClient {
    let keyId: String
    let teamId: String
    let bundleId: String
    private let session: URLSession
    private var cachedJWT: CachedJWT?
    /// Stored once at init so we don't re-read Keychain on every send.
    private let privateKey: P256.Signing.PrivateKey?

    /// Throws APNsError.invalidKey if the stored .p8 can't be parsed.
    /// Throws APNsError.missingKey if no key is stored yet.
    init(keyId: String, teamId: String, bundleId: String, session: URLSession = .shared) throws {
        self.keyId = keyId
        self.teamId = teamId
        self.bundleId = bundleId
        self.session = session

        guard let pemData = APNsKeyStore.get() else {
            self.privateKey = nil
            throw APNsError.missingKey
        }
        guard let pem = String(data: pemData, encoding: .utf8) else {
            self.privateKey = nil
            throw APNsError.invalidKey("key bytes are not valid UTF-8")
        }
        do {
            self.privateKey = try P256.Signing.PrivateKey(pemRepresentation: pem)
        } catch {
            self.privateKey = nil
            throw APNsError.invalidKey("CryptoKit couldn't parse PEM: \(error.localizedDescription)")
        }
    }

    /// Build a fresh JWT. Exposed as `internal` so tests can verify the
    /// signature shape against the public key. Callers should prefer
    /// `send(...)` which handles caching.
    func makeJWT(now: Date = Date()) throws -> String {
        guard let privateKey else { throw APNsError.missingKey }

        // Header: {"alg":"ES256","kid":"<keyId>"}
        let header: [String: String] = ["alg": "ES256", "kid": keyId]
        // Payload: {"iss":"<teamId>","iat":<seconds-since-epoch>}
        let payload: [String: Any] = ["iss": teamId, "iat": Int(now.timeIntervalSince1970)]

        guard let headerData = try? JSONSerialization.data(withJSONObject: header, options: [.sortedKeys]),
              let payloadData = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]) else {
            throw APNsError.invalidKey("could not serialize JWT header/payload")
        }

        let headerB64 = Self.base64URLEncode(headerData)
        let payloadB64 = Self.base64URLEncode(payloadData)
        let signingInput = "\(headerB64).\(payloadB64)"
        guard let signingInputData = signingInput.data(using: .utf8) else {
            throw APNsError.invalidKey("ASCII encode failed")
        }
        let signature = try privateKey.signature(for: signingInputData)
        // ECDSA signature is 64 bytes raw (r || s); that's the JWS/JWT
        // expected form. CryptoKit's `.rawRepresentation` gives exactly that.
        let sigB64 = Self.base64URLEncode(signature.rawRepresentation)
        return "\(signingInput).\(sigB64)"
    }

    /// Base64url per RFC 7515: standard base64, minus trailing padding,
    /// with `+`→`-` and `/`→`_`.
    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private func currentJWT() throws -> String {
        if let cached = cachedJWT, !cached.isExpired {
            return cached.token
        }
        let fresh = try makeJWT()
        cachedJWT = CachedJWT(token: fresh, issuedAt: Date())
        return fresh
    }

    /// Send a pre-encoded JSON alert payload to a single device. On
    /// APNs 410 / Unregistered / BadDeviceToken, throws `.unregistered`
    /// so the caller can drop the device from storage. On 429/503,
    /// retries ONCE with a 2s backoff before surfacing `.throttled`.
    ///
    /// Payload is passed as Data (not [String: Any]) so it crosses
    /// concurrency domains cleanly — [String: Any] isn't Sendable in
    /// Swift 6 strict mode.
    func send(payloadData body: Data, toDevice device: RegisteredPushDevice) async throws {
        let host = device.environment == "production"
            ? "api.push.apple.com"
            : "api.sandbox.push.apple.com"
        guard let url = URL(string: "https://\(host)/3/device/\(device.token)") else {
            throw APNsError.badRequest("invalid device token format")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("alert", forHTTPHeaderField: "apns-push-type")
        request.setValue(bundleId, forHTTPHeaderField: "apns-topic")
        request.setValue("10", forHTTPHeaderField: "apns-priority")
        request.setValue("bearer \(try currentJWT())", forHTTPHeaderField: "authorization")

        do {
            try await performSend(request: request, isRetry: false)
        } catch APNsError.throttled {
            // One retry with a small backoff. If the second attempt
            // also gets throttled, surface the error — caller decides
            // whether to show it in the UI or silently drop.
            try await Task.sleep(nanoseconds: 2_000_000_000)
            try await performSend(request: request, isRetry: true)
        }
    }

    private func performSend(request: URLRequest, isRetry: Bool) async throws {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APNsError.unknown(-1, "non-HTTP response")
        }
        switch http.statusCode {
        case 200:
            return
        case 400:
            let reason = Self.extractReason(from: data)
            if reason == "BadDeviceToken" || reason == "DeviceTokenNotForTopic" || reason == "Unregistered" {
                throw APNsError.unregistered
            }
            throw APNsError.badRequest(reason)
        case 403:
            // Usually InvalidProviderToken (bad JWT). Surface so the
            // Settings UI can prompt the user to re-upload the .p8.
            let reason = Self.extractReason(from: data)
            throw APNsError.invalidKey(reason)
        case 410:
            throw APNsError.unregistered
        case 429, 503:
            if isRetry { throw APNsError.throttled }
            throw APNsError.throttled
        case 500...599:
            throw APNsError.serverError(http.statusCode)
        default:
            let reason = Self.extractReason(from: data)
            throw APNsError.unknown(http.statusCode, reason)
        }
    }

    /// APNs error bodies are `{"reason":"<string>"}`. Extract it or
    /// return an empty string so the error message isn't just garbage.
    private static func extractReason(from data: Data) -> String {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let reason = obj["reason"] as? String else { return "" }
        return reason
    }
}

File: QuipLinux/src/services/apns_client.rs
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
use serde::{Deserialize, Serialize};

use super::apns_key_store::ApnsKeyStore;

/// Errors surfaced from `ApnsClient::send`. Mirrors Mac's `APNsError` cases.
/// Distinct cases let the caller respond differently — `Unregistered` →
/// drop device from store, `Throttled` → retry, `BadKey` → surface in UI.
#[derive(Debug, Clone)]
pub enum ApnsError {
    MissingKey,
    InvalidKey(String),
    BadRequest(String),
    Unregistered,
    Throttled,
    ServerError(u16),
    Unknown(u16, String),
}

impl std::fmt::Display for ApnsError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::MissingKey => write!(f, "no APNs key configured"),
            Self::InvalidKey(m) => write!(f, "invalid APNs key: {m}"),
            Self::BadRequest(m) => write!(f, "APNs bad request: {m}"),
            Self::Unregistered => write!(f, "APNs reports device unregistered"),
            Self::Throttled => write!(f, "APNs throttled"),
            Self::ServerError(s) => write!(f, "APNs server error {s}"),
            Self::Unknown(s, r) => write!(f, "APNs {s}: {r}"),
        }
    }
}

impl std::error::Error for ApnsError {}

/// Cached JWT — APNs rejects tokens older than 60 minutes, so we rotate at
/// 50 min to leave headroom (matching Mac's `CachedJWT.isExpired`).
struct CachedJwt {
    token: String,
    issued_at: Instant,
}

impl CachedJwt {
    fn is_expired(&self) -> bool {
        self.issued_at.elapsed() > Duration::from_secs(50 * 60)
    }
}

#[derive(Debug, Clone, Serialize)]
struct JwtClaims {
    iss: String,
    iat: i64,
}

/// Per-device record. Mirrors Mac's `RegisteredPushDevice`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RegisteredPushDevice {
    pub token: String,
    pub environment: String, // "development" | "production"
    pub registered_at: u64,  // unix seconds
}

/// Signs ES256 JWTs with the .p8 key and POSTs them + an alert payload to
/// Apple's HTTP/2 APNs endpoint via reqwest's HTTP/2 client. Mirrors
/// `QuipMac/Services/APNsClient.swift`.
pub struct ApnsClient {
    pub key_id: String,
    pub team_id: String,
    pub bundle_id: String,
    encoding_key: EncodingKey,
    cached_jwt: Mutex<Option<CachedJwt>>,
    http: reqwest::blocking::Client,
}

impl ApnsClient {
    pub fn new(key_id: String, team_id: String, bundle_id: String) -> Result<Arc<Self>, ApnsError> {
        let pem = ApnsKeyStore::get().ok_or(ApnsError::MissingKey)?;
        Self::new_from_pem(&pem, key_id, team_id, bundle_id)
    }

    /// Construct from an explicit PEM byte slice instead of the configured
    /// keystore. Lets tests use a throwaway key without touching the user's
    /// config dir.
    pub fn new_from_pem(
        pem: &[u8],
        key_id: String,
        team_id: String,
        bundle_id: String,
    ) -> Result<Arc<Self>, ApnsError> {
        let encoding_key = EncodingKey::from_ec_pem(pem)
            .map_err(|e| ApnsError::InvalidKey(format!("PEM parse failed: {e}")))?;

        // HTTP/2 is required by Apple's APNs gateway. reqwest supports it
        // when `http2` is in the feature list (which we set in Cargo.toml).
        let http = reqwest::blocking::Client::builder()
            .http2_prior_knowledge()
            .timeout(Duration::from_secs(15))
            .build()
            .map_err(|e| ApnsError::Unknown(0, format!("HTTP client init failed: {e}")))?;

        Ok(Arc::new(Self {
            key_id,
            team_id,
            bundle_id,
            encoding_key,
            cached_jwt: Mutex::new(None),
            http,
        }))
    }

    /// Build a fresh JWT. Public so tests can inspect the structure.
    pub fn make_jwt(&self) -> Result<String, ApnsError> {
        let mut header = Header::new(Algorithm::ES256);
        header.kid = Some(self.key_id.clone());
        let claims = JwtClaims {
            iss: self.team_id.clone(),
            iat: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs() as i64)
                .unwrap_or(0),
        };
        encode(&header, &claims, &self.encoding_key)
            .map_err(|e| ApnsError::InvalidKey(format!("JWT sign failed: {e}")))
    }

    fn current_jwt(&self) -> Result<String, ApnsError> {
        let mut guard = self.cached_jwt.lock().expect("apns jwt poisoned");
        if let Some(c) = guard.as_ref() {
            if !c.is_expired() {
                return Ok(c.token.clone());
            }
        }
        let fresh = self.make_jwt()?;
        *guard = Some(CachedJwt {
            token: fresh.clone(),
            issued_at: Instant::now(),
        });
        Ok(fresh)
    }

    /// POST a pre-encoded JSON alert payload to a single device.
    /// `payload_json` should already be the JSON object body; we don't
    /// touch it. `collapse_id` truncated to 64 bytes per APNs spec.
    /// On 410/Unregistered/BadDeviceToken the caller should drop the
    /// device from its registry — `Err(Unregistered)` signals that.
    /// 429/503 retries once with a 2s sleep, then surfaces `Throttled`.
    pub fn send(
        &self,
        payload_json: &str,
        device: &RegisteredPushDevice,
        collapse_id: Option<&str>,
    ) -> Result<(), ApnsError> {
        match self.do_send(payload_json, device, collapse_id) {
            Err(ApnsError::Throttled) => {
                std::thread::sleep(Duration::from_secs(2));
                self.do_send(payload_json, device, collapse_id)
            }
            other => other,
        }
    }

    fn do_send(
        &self,
        payload_json: &str,
        device: &RegisteredPushDevice,
        collapse_id: Option<&str>,
    ) -> Result<(), ApnsError> {
        let host = if device.environment == "production" {
            "api.push.apple.com"
        } else {
            "api.sandbox.push.apple.com"
        };
        let url = format!("https://{host}/3/device/{}", device.token);

        let mut req = self
            .http
            .post(&url)
            .header("apns-push-type", "alert")
            .header("apns-topic", &self.bundle_id)
            .header("apns-priority", "10")
            .header("authorization", format!("bearer {}", self.current_jwt()?))
            .header("content-type", "application/json")
            .body(payload_json.to_string());

        if let Some(c) = collapse_id.filter(|s| !s.is_empty()) {
            // 64-byte cap per APNs spec.
            let truncated = if c.len() > 64 { &c[..64] } else { c };
            req = req.header("apns-collapse-id", truncated);
        }

        let resp = req.send().map_err(|e| {
            // Network-level failure (TLS, DNS, etc.). Caller may want to
            // retry or surface to the user; treat as Unknown for parity
            // with Mac's NSURLErrorDomain catch-all.
            ApnsError::Unknown(0, format!("send failed: {e}"))
        })?;

        let status = resp.status().as_u16();
        let body = resp.text().unwrap_or_default();
        let reason = extract_reason(&body);

        match status {
            200 => Ok(()),
            400 => {
                if reason == "BadDeviceToken"
                    || reason == "DeviceTokenNotForTopic"
                    || reason == "Unregistered"
                {
                    Err(ApnsError::Unregistered)
                } else {
                    Err(ApnsError::BadRequest(reason))
                }
            }
            403 => Err(ApnsError::InvalidKey(reason)),
            410 => Err(ApnsError::Unregistered),
            429 | 503 => Err(ApnsError::Throttled),
            500..=599 => Err(ApnsError::ServerError(status)),
            other => Err(ApnsError::Unknown(other, reason)),
        }
    }
}

fn extract_reason(body: &str) -> String {
    #[derive(Deserialize)]
    struct Body {
        reason: Option<String>,
    }
    serde_json::from_str::<Body>(body)
        .ok()
        .and_then(|b| b.reason)
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    /// JWT signing requires a real EC P-256 key, which we'd have to embed
    /// or generate; either way the unit test would just be exercising the
    /// `jsonwebtoken` crate. The error paths below cover the failure
    /// shapes a wrong .p8 file produces.

    #[test]
    fn extract_reason_parses_apns_error_body() {
        assert_eq!(extract_reason(r#"{"reason":"BadDeviceToken"}"#), "BadDeviceToken");
        assert_eq!(extract_reason("not json"), "");
        assert_eq!(extract_reason(""), "");
    }

    #[test]
    fn invalid_pem_surfaces_clear_error() {
        let r = ApnsClient::new_from_pem(b"not a real PEM", "k".into(), "t".into(), "b".into());
        assert!(matches!(r, Err(ApnsError::InvalidKey(_))));
    }
}
